File: core/src/commonMain/kotlin/io/umain/munchies/core/di/ScopeExtensions.kt
package io.umain.munchies.core.di

import org.koin.core.scope.Scope

/**
 * KMP helper to retrieve ViewModels from a Koin Scope.
 *
 * On Android: Calls scope.get<T>() directly.
 * On iOS: Provides a stable interop point for Swift to retrieve typed instances.
 *
 * Usage from Swift:
 * ```swift
 * let viewModel = try! scope.get(viewModel: RestaurantDetailViewModel.self)
 * ```
 *
 * Usage from Kotlin:
 * ```kotlin
 * val viewModel = scope.getViewModel<RestaurantDetailViewModel>()
 * ```
 */
inline fun <reified T> Scope.getViewModel(): T = get()

/**
 * KMP helper to retrieve ViewModels from a Koin Scope with parameters.
 *
 * On Android: Calls scope.get<T>(parameters = { ... }) directly.
 * On iOS: Provides a stable interop point for Swift with parameter passing.
 *
 * Usage from Kotlin:
 * ```kotlin
 * val viewModel = scope.getViewModel<RestaurantDetailViewModel> {
 *     parametersOf(restaurantId)
 * }
 * ```
 */
inline fun <reified T> Scope.getViewModel(noinline parameterBuilder: () -> org.koin.core.parameter.ParametersHolder): T =
    get(parameters = parameterBuilder)

File: iosApp/iosApp/Navigation/ScopeExtensions.swift
import Foundation
import shared

/// Swift extension to retrieve typed ViewModels from Koin Scope.
///
/// Koin's Scope class is defined in Kotlin, but Swift cannot call generic functions
/// with reified type parameters directly. This extension provides type-safe syntax
/// by bridging to the non-generic Kotlin helpers.
extension Scope {
    /// Retrieve a ViewModel from this scope by type.
    ///
    /// Usage:
    /// ```swift
    /// let viewModel = try! scope.get(viewModel: RestaurantDetailViewModel.self)
    /// ```
    func get<T>(viewModel type: T.Type) -> T {
        let kotlinType = type as! AnyClass
        let vmObject = self.getViewModelTyped(type: kotlinType)
        return vmObject as! T
    }
    
    /// Retrieve a ViewModel from this scope with parameters.
    ///
    /// Usage:
    /// ```swift
    /// let viewModel = try! scope.get(
    ///     viewModel: RestaurantDetailViewModel.self,
    ///     parameters: restaurantId
    /// )
    /// ```
    func get<T>(viewModel type: T.Type, parameters: Any...) -> T {
        let kotlinType = type as! AnyClass
        let vmObject = self.getViewModelTyped(type: kotlinType, parameters: parameters)
        return vmObject as! T
    }
}
